content/docs/user-guide/dvc-files.md

```diff
@@ -237,14 +242,224 @@ stages:
+and `outs` entries of `.dvc` files).
+
+Full <abbr>parameters</abbr> (key and value) are listed too (under `params`),
+grouped by parameters file. And in the case of templated/parameterized
+`dvc.yaml` files (see next section), their actual values are substituted into
+the `dvc.lock` YAML structure.
+
+### Parameterize `dvc.yaml`
+
+`dvc.yaml` supports a special templating format that enables reusing simple
+values from different sources in the YAML structure itself.
+
+Global parameters (file scope) can be defined either in a
+[params file](/doc/command-reference/params) (`params.yaml` by default), or in a
+top `vars` section of `dvc.yaml`. Both can coexist as long as `vars` don't
+overwrite any external param.
+
+Let's say we have a `params.yaml` file with the following contents:
+
+```yaml
+models:
+  us:
+    threshold: 10
+    filename: 'model-us.hdf5'
+```
+
+To use those values in `dvc.yaml`, you can use the `${}` expression anywhere in
+the file:
+
+```yaml
+stages:
+  build-us:
+    cmd: >-
+      python train.py
+      --tresh ${models.us.threshold}
+      --out ${models.us.filename}
+    outs:
+      - ${models.us.filename}:
+        cache: true
+```
+
+Alternatively, the same or other values can be included as `vars` like this:
+
+```yaml
+vars:
+  - models:
+    us:
+      thresh: 10
+  - desc: 'Reusable description'
+
+stages:
+  build-us:
+    desc: ${desc}
+    cmd: python train.py --tresh ${models.us.threshold} ...
+```
+
+The global `vars` section also accepts any number of parameters file names as
+entries as a way to include their contents for substituting in `dvc.yaml`, e.g.:
+
+```yaml
+vars:
+  - params.json
+  - myconfig.yaml
+  - myvar: value
+```
+
+> ⚠️ Note that the default `params.yaml` is always included (completely) first.
+
+It's also possible to narrow down what to include from each params file by using
+this special `:` colon format:
+
+```yaml
+vars:
+  - params.json:cleaning,featurization
+
+stages:
+  clean:
+    cmd: source ${cleaning.script}
+    outs:
+      - ${cleaning.outname}
+  featurize:
+    cmd: python ${featurization.pyscript}
+    deps:
+      - ${cleaning.outname}
+    outs: ...
+```
+
+> DVC merges structures from params files as long as there are no collisions.
+> E.g. `{"grp": {"a": 1}}` can be merged with `{"grp": {"b": 2}}` but not with
+> `{"grp": {"a": 10}}`.
+
+Local parameters (stage scope) are also supported, with stage-specific `vars`
+sections:
+
+```yaml
+stages:
+  build-us:
+    vars:
+      - model:
+        filename: 'model-us.hdf5'
+    cmd: python train.py --out ${model.filename}
+    outs:
+      - ${model.filename}
+```
+
+The template substitution expression (`${}`) supports these forms:
+
+```yaml
+${param} # Simple
+${param.key} # Nested values through . (period)
+${param.list[0]} # List elements via index in [] (square brackets)
+```
+
+> To use the same expression literally in `dvc.yaml`, escape it with a
+> backslash, e.g. `\${...`.
+
+### Stage groups (loops)
```

Comment:
groups or loops? or something that would user expects?

looks like we are not sure on the terminology

Reply:
Better to just say `foreach` here.

Reply:
But what's the point of these "loops"? To generate stage groups. They get expanded in dvc.lock. I think that is the more relevant concept/term, but the "loop" part is an important detail on how to use them.

I kept both in the section title for SEO purposes but I don't have a strong opinion, I'm happy to remove "loops" from there, lmk.

Reply:
To clarify my position, here's an alternative (equally valid to my mind) way in which the syntax could've been implemented without the "foreach" term (which people may associate with "loops"):

```yaml
stages:
  mygroup:
    set:
      - foo
      - bar
    cmd: echo ${foo}
    ...
```

In the above example is clearer that the looping part is secondary, I think. Could be called "iterating over the set" instead.

> p.s. something related to this is mentioned in iterative/dvc/issues/5165

Reply:
> Better to just say foreach or foreach loops here.

it depends on what to we expect from this doc - is it a language reference? then we should name sections with keywords. But we'll need a proper document which will connect dots and explain these new concepts from the user perspective.

We've been implementing this based on a lot of tickets - how did users ask about this? what terminology they were using?

> To generate stage groups

we don't have a term `group` defined. When you introduce it if might feel that we have some special meaning to it, while it's not (there is no special concept of a group on the dvc.lock file level)

word `generate` might be a good bu itself though

Reply:
> is it a language reference? then we should name sections with keywords

BTW the `foreach` and `do` sections are named of course.

> we don't have a term group defined. When you introduce it if might feel that we have some special meaning

It's a very common word. A "group of stages" seems self-explanatory to me...?

Good idea about "generate". Introduced in 70b4b9a6.

Reply:
Continued in https://github.com/iterative/dvc.org/pull/2052#pullrequestreview-559754434. Resolving this one.